fdbclient/include/fdbclient/MetaclusterManagement.actor.h

```diff
@@ -1062,6 +1055,315 @@ Future<Void> managementClusterRemoveTenantFromGroup(Transaction tr,
+			                      Optional<ClusterConnectionString>(self->connectionString),
+			                      updatedEntry);
+		}
+
+		TraceEvent("MarkedDataClusterReady")
+		    .detail("Name", self->ctx.clusterName.get())
+		    .detail("Version", tr->getCommittedVersion());
+
+		return Void();
+	}
+
+	ACTOR static Future<Void> markManagementTenantAsError(RestoreClusterImpl* self,
+	                                                      Reference<typename DB::TransactionT> tr) {
+		state Optional<TenantMapEntry> tenantEntry = wait(tryGetTenantTransaction(tr, self->tenantName));
+
+		if (!tenantEntry.present()) {
+			return Void();
+		}
+
+		tenantEntry.get().tenantState = TenantState::ERROR;
+		ManagementClusterMetadata::tenantMetadata().tenantMap.set(tr, self->tenantName, tenantEntry.get());
+		return Void();
+	}
+
+	ACTOR static Future<bool> getTenantsFromDataCluster(RestoreClusterImpl* self, Reference<ITransaction> tr) {
+		TenantNameRef begin = self->lastTenantName;
+		TenantNameRef end = "\xff\xff"_sr;
+		state Future<KeyBackedRangeResult<std::pair<TenantName, TenantMapEntry>>> tenantsFuture =
+		    TenantMetadata::tenantMap().getRange(tr, begin, end, CLIENT_KNOBS->MAX_TENANTS_PER_CLUSTER);
+		state KeyBackedRangeResult<std::pair<TenantName, TenantMapEntry>> tenants = wait(tenantsFuture);
+
+		if (!tenants.results.empty()) {
+			for (auto t : tenants.results) {
+				self->dataClusterTenantMap.emplace(t.first, t.second);
+				self->lastTenantName = t.first;
+			}
+		}
+
+		return !tenants.more;
+	}
+
+	ACTOR static Future<Void> getAllTenantsFromDataCluster(RestoreClusterImpl* self) {
+		loop {
+			bool gotAllTenants = wait(self->ctx.runDataClusterTransaction(
+			    [self = self](Reference<ITransaction> tr) { return getTenantsFromDataCluster(self, tr); }));
+
+			if (gotAllTenants) {
+				break;
+			}
+		}
+
+		return Void();
+	}
+
+	ACTOR static Future<bool> getTenantsFromManagementCluster(RestoreClusterImpl* self, Reference<ITransaction> tr) {
+		TenantNameRef begin = self->lastTenantName;
+		TenantNameRef end = "\xff\xff"_sr;
+		state Future<KeyBackedRangeResult<std::pair<TenantName, TenantMapEntry>>> tenantsFuture =
+		    ManagementClusterMetadata::tenantMetadata().tenantMap.getRange(
+		        tr, begin, end, CLIENT_KNOBS->MAX_TENANTS_PER_CLUSTER);
+		state KeyBackedRangeResult<std::pair<TenantName, TenantMapEntry>> tenants = wait(tenantsFuture);
+
+		if (!tenants.results.empty()) {
+			for (auto t : tenants.results) {
+				self->mgmtClusterTenantMap.emplace(t.first, t.second);
+				self->lastTenantName = t.first;
+			}
+		}
+
+		return !tenants.more;
+	}
+
+	ACTOR static Future<bool> getTenantsFromManagementClusterForCurrentDataCluster(RestoreClusterImpl* self,
+	                                                                               Reference<ITransaction> tr) {
+		std::pair<Tuple, Tuple> clusterTupleRange =
+		    std::make_pair(Tuple::makeTuple(self->lastClusterName), Tuple::makeTuple(keyAfter(self->clusterName)));
```

Comment:
I don't think we need a `lastClusterName` concept here, we can just read the entries for `clusterName`. And as stated for the data cluster case, we can probably assume this happens in a single transaction given our constraints on the number of tenants per cluster.